torch/csrc/api/src/optim/optimizer.cpp

```diff
@@ -29,10 +29,9 @@ void OptimizerBase::add_parameters(const ParameterCursor& cursor) {
 
 void OptimizerBase::zero_grad() {
   for (auto& parameter : parameters_) {
-    auto& grad = parameter.grad();
-    if (grad.defined()) {
-      grad = grad.detach();
-      Tensor(grad).data().zero_();
+    if (parameter.grad().defined()) {
+      parameter.grad().detach_();
+      parameter.grad().zero_();
```

Comment:
I wonder if this should also be `detach().zero()`? Or if not, could you add a test like:

```
auto model = Something();
auto optim = SGD(model->parameters(), 1e-3);
optim->zero_grad();
auto random_operation = model->parameters()->begin()->value.grad() * 2;
```

My hypothesis is that the grad() will point the one before the zero_, and then the code that detects if an in-place op was done will run and it will runtime_error.

Reply:
Alternatively, have a NoGradGuard here as well.

Reply:
I thought double backwards was possible? In which case the grad might be a node in an autograd graph, no?